Rework raiders and fix raider issues.

someaddons opened the pull request:
Closes #
Closes #
Closes #

# Changes proposed in this pull request:
Originally this was to fix loading/unloading/counting issues with the raiders, but ended up reworking it a bit and creating a new event system which can be used for future colony events as well.

Whats new: We now have ColonyEvents, which are registered through a forge registry for them, so other mods can add events aswell. Colonyevents are managed through a new EventManager, which updates them and saves/reads them from nbt. Those events receive a regular tick from the colony,albeit pretty slow(each 500 ticks) to update their status and do whatever update logic they need to. Colony events can also register/unregister entities related to them and spawn temporary structures. For managing the structures the EventManager has an EventStructureManager. The structure manager allows spawning structures for a given event and takes a backup schematic of the surroindings. Once the event is removed(or was not loaded during colony load) the backup schematic is put back in place.

I've translated both the raider and pirate events over as colonyevents, which also allows us now to spawn multiple ones seperatly if we want to. Atm I've added that once a raid succeeds the biggest raid size it gets split up into an additional raid from another spawn pos. 

Both raid events now keep track of their entities and respawn them if those were not actually killed(onDeath triggered), but removed by other means, e.g. minecraft itself removing it, this also includes chunk unloads. The respawn is limited to the previous position, or a max distance of the original spawn position, so that when the colony is still active, but the raiders happen to be unloaded it tries to place them a bit closer in a loaded chunk.

The PirateEvent now has a maximum amount of pirates it can have active, depending on the amount of spawners remaining(spawnercount * 2 +10). It also now spawns additional pirates,depending on shipsize, without a player beeing close to the spawners, so that it always raids the colony.

The BarbarianEvent has a given Horde(as previously) which has a predetermined size of archers/melees/bosses. It actively keeps track of all and respawns any missing. Any additional mobs which try to register over the Horde's counter are killed off, this e.g. might happen with chunk unloading/loading.

Barracks got a little new feature, now that we can have more than one raid from different angles, which gives the raid mobs a glow effect so they are easier to find. This is called "Spies" and uses gold to hire those spies  into the enemy lines which then lets the players know the position.

Review please

someaddons committed:
Rework raiders and fix raider issues.

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java  5
- src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianHorde.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java  8
- src/main/java/com/minecolonies/coremod/network/messages/HireSpiesMessage.java  4
- src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java  8
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateRaidEvent.java  10
- src/main/java/com/minecolonies/coremod/colony/managers/EventStructureManager.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/ShipSize.java  4
- src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventTypeRegistryEntry.java  2
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianEventUtils.java  4
- src/main/java/com/minecolonies/apiimp/initializer/ModColonyEventTypeInitializer.java  2
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateEventUtils.java  8
         

Clones added
============
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4998947&bid=16496596)

Raycoms commented on src/api/java/com/minecolonies/api/colony/colonyEvents/IColonyEvent.java:
```diff
@@ -0,0 +1,136 @@
+    default public void unregisterEntity(final Entity entity) {}
+
+    ;
+
```
formatting seems off here

someaddons replied:
just slightly xD

Raycoms commented on src/api/java/com/minecolonies/api/colony/IColonyView.java:
```diff
@@ -463,4 +454,6 @@
      */
     List<CompactColonyReference> getFeuds();
+
+    boolean isSpiesEnabled();
```
not areSpiesEnabled ?

someaddons replied:
ye

Raycoms commented on src/api/java/com/minecolonies/api/entity/citizen/AbstractEntityCitizen.java:
```diff
@@ -115,10 +115,9 @@ public int getTicksExisted()
     {
-        return null;
+        return this;
     }
```
We had null here on purpose to avoid them changing dimensions, is that wrong?

someaddons replied:
ah no don't think so, only wanted to change the overwrite to the right method

Raycoms commented on src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java:
```diff
@@ -75,6 +71,18 @@
+     */
+    private boolean isRegistered = false;
+
+    private int invulTime = 2 * 20;
```
jdoc, what is this?

Raycoms commented on src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java:
```diff
@@ -329,7 +366,14 @@ public IColony getColony()
     public void registerWithColony()
     {
-        getColony();
+        if (colony == null || eventID == 0)
```
Good idea

someaddons replied:
ye all raiders now have to know their colony and eventID and are removed otherwise. The events gonna take care to spawn a new one if needed for the event

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java:
```diff
@@ -0,0 +1,104 @@
+        }
+        findPaneOfTypeByID(TEXT_ID, Text.class).setTextContent(
+          "During raids you can hire spies here, to infiltrate the enemy lines. Those then let you know enemy positions although they demand some gold as payment. Put the gold into the barracks inventory to automatically pay them.");
+    }
```
lang entry?

someaddons replied:
ye missing^^

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -149,9 +149,9 @@
+     * Wether the colony is raided
      */
-    private int horde;
+    private boolean isRaided;
```
isUnderRaid, isBeingRaided might be more understandable

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java:
```diff
@@ -0,0 +1,399 @@
+        }
+
+        archers.remove(entity);
+        chiefs.remove(entity);
```
why in this one here we just remove it from all and in the other ones we do it in the instanceof check?

someaddons replied:
ye here its only needed to remove the entity from the list, while the other cases have special type-dependent logic, like reducing archer count, or when adding to only add to the right list

someaddons committed:
cleanup

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/HireSpiesMessage.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianHorde.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateEventUtils.java  8
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java  8
- src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java  5
- src/main/java/com/minecolonies/apiimp/initializer/ModColonyEventTypeInitializer.java  2
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/ShipSize.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateRaidEvent.java  10
- src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java  8
- src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java  4
- src/main/java/com/minecolonies/coremod/colony/managers/EventStructureManager.java  4
- src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventTypeRegistryEntry.java  2
         

Clones added
============
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4998947&bid=16496596)

someaddons committed:
rebalance raid sizes and improve stuck handling

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateRaidEvent.java  10
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianHorde.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/ShipSize.java  4
- src/main/java/com/minecolonies/coremod/colony/managers/EventStructureManager.java  4
- src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java  5
- src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java  4
- src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIBreakDoor.java  1
- src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java  8
- src/main/java/com/minecolonies/apiimp/initializer/ModColonyEventTypeInitializer.java  2
- src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventTypeRegistryEntry.java  2
- src/main/java/com/minecolonies/coremod/network/messages/HireSpiesMessage.java  4
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateEventUtils.java  8
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java  13
         

Clones added
============
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4998947&bid=16496596)

someaddons committed:
Merge remote-tracking branch 'remotes/origin/version/1.12' into raiderFix112

# Conflicts:
#	build.properties

someaddons committed:
fix pathing error

Raycoms commented on src/api/java/com/minecolonies/api/colony/colonyEvents/IColonyEvent.java:
```diff
@@ -0,0 +1,146 @@
+     */
+    default public void registerEntity(final Entity entity) {}
+
+    ;
```
those are still off, the semicolons

Raycoms commented on src/api/java/com/minecolonies/api/entity/citizen/AbstractEntityCitizen.java:
```diff
@@ -115,8 +115,7 @@ public int getTicksExisted()
     @Override
-    @Nullable
-    public Entity changeDimension(final int dimensionIn)
+    public Entity changeDimension(int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
```
Can't we import the ITeleporter for this?

someaddons replied:
ye we can, just had that copied

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java:
```diff
@@ -0,0 +1,426 @@
+        }
+
+
+            for (final Entity entity : getEntities())
```
formatting here too

someaddons replied:
oh and the debug logs are still in, those arent needed

someaddons committed:
Finish walking AI and some cleanup

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/client/gui/WindowBarracksBuilding.java  1
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/ShipSize.java  4
- src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java  8
- src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIWalkToRandomHuts.java  6
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianHorde.java  4
- src/main/java/com/minecolonies/coremod/colony/ColonyManagerWorldAccess.java  1
- src/main/java/com/minecolonies/apiimp/initializer/ModColonyEventTypeInitializer.java  2
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  6
- src/main/java/com/minecolonies/coremod/colony/managers/EventStructureManager.java  4
- src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventTypeRegistryEntry.java  2
- src/main/java/com/minecolonies/coremod/colony/managers/RaidManager.java  4
- src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java  2
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java  9
- src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java  5
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateRaidEvent.java  10
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateEventUtils.java  8
- src/main/java/com/minecolonies/coremod/network/messages/HireSpiesMessage.java  4
- src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIBreakDoor.java  1
- src/main/java/com/minecolonies/coremod/network/messages/BuyCitizenMessage.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java  4
         

Complexity decreasing per file
==============================
+ src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIRaiderAttackMelee.java  -1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4998947&bid=16496596)

marchermans commented on src/api/java/com/minecolonies/api/colony/colonyEvents/IColonyEvent.java:
```diff
@@ -0,0 +1,142 @@
+/**
+ * Interface for colony event types.
+ */
+public interface IColonyEvent
```
IColonyEvent is named a bit weirdly, especially since you force it to have a spawnpoint.
Maybe think of something that more allines with your goals.

marchermans replied:
Something like `IColonyEntitySpawnEvent` or `IColonyGroupSpawnEvent` possibly with some abstraction into an `IColonySpawnEvent` and `IColonyEvent`

marchermans replied:
Also you might want to have an interface that describes an event that spawns a structure.
So that raid events are both spawning and have structure handling.

This makes expanding this system easier in the future -> Think of camps of traiders that spawn outside the city walls to trade with etc.

someaddons replied:
ye you're right indeed, colony event doesnt necessarily need entities, splitting it makes sense. An interface for spawning structures alongside an event should be useful too

someaddons replied:
I've seperated them into Colony/Spawn/entitySpawn/raid/structureSpawn types now, think that should fit it

marchermans reviewed (changes requested):
The general idea here is perfect.
Nice work.

However i think some minor rework on what actually defines an Event is the only thing that is needed.
In my eyes `IColonyEvent` is not linked to an entity. But only to a colony.

Now sub classes of this event might relate to entity spawn, group spawning of entities (packspawning in MC terms) or even the spawning of a structure. But these events have initially nothing in common with normal events and their logic should be split up.

See if you can rework your event handling to be more geared towards future expansion of this system.

someaddons committed:
Abstract Interfaces further
Remove structure backup file after loading backup
Mourning now has lower priority than sleeping/hiding
Improved raider unstuck a bit further

someaddons committed:
Merge remote-tracking branch 'remotes/origin/version/1.12' into raiderFix112

# Conflicts:
#	build.properties

marvin-bitterlich commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/apiimp/initializer/ModColonyEventTypeInitializer.java  2
- src/main/java/com/minecolonies/coremod/colony/managers/RaidManager.java  4
- src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventTypeRegistryEntry.java  2
- src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java  2
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianHorde.java  4
- src/main/java/com/minecolonies/coremod/colony/managers/EventStructureManager.java  4
- src/main/java/com/minecolonies/coremod/client/gui/WindowBarracksBuilding.java  1
- src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIWalkToRandomHuts.java  6
- src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java  5
- src/main/java/com/minecolonies/coremod/network/messages/BuyCitizenMessage.java  1
- src/api/java/com/minecolonies/api/entity/mobs/RaiderMobUtils.java  8
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/ShipSize.java  4
- src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIBreakDoor.java  1
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  6
- src/main/java/com/minecolonies/coremod/network/messages/HireSpiesMessage.java  4
- src/main/java/com/minecolonies/coremod/client/gui/WindowsBarracksSpies.java  4
- src/main/java/com/minecolonies/coremod/colony/ColonyManagerWorldAccess.java  1
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateRaidEvent.java  10
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/pirateEvent/PirateEventUtils.java  8
- src/main/java/com/minecolonies/coremod/colony/colonyEvents/raidEvents/babarianEvent/BarbarianRaidEvent.java  9
         

Complexity decreasing per file
==============================
+ src/main/java/com/minecolonies/coremod/entity/mobs/aitasks/EntityAIRaiderAttackMelee.java  -1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/commands/killcommands/RaiderKillCommand.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4998947&bid=16496596)
